[WIP] Try building Joinfaces with Gradle

larsgrefer opened the pull request.

larsgrefer commented:
@persapiens What's your opinion on gradle?

persapiens commented:
Some questions:

- joinfaces-dependencies subproject does not exist anymore. Is it not useful to maven users like illustrated [here](https://github.com/spring-projects/spring-boot/wiki/Building-On-Spring-Boot#dependency-management)?
- I like the idea of using properties to ease library versions upgrade. Is it possible to use something like this:
```groovy
ext {
  primefacesVersion = '6.2'
  butterfacesVersion = '3.0.6'
}
```
- Library versions are written across `build.gradle` files now. More, some versions are duplicated like primefaces-extensions: main build.gradle and starters/primefaces-spring-boot-starter/build.gradle. How could we handle it better?
- I didn't see [cpd](https://plugins.gradle.org/plugin/de.aaschmid.cpd) check.
- I didn't see duplicate classes check. We could use [nebula-lint](https://plugins.gradle.org/plugin/nebula.lint) with:
```groovy
gradleLint {
  rules = ['transitive-duplicate-dependency-class']
  criticalRules = ['duplicate-dependency-class','dependency-parentheses']
}
```
- I didn't see cycles check. We could use jdepend with:
```groovy
jdependMain.doLast {
  File fileMain = new File(jdepend.reportsDir, "main.xml");
  assert fileMain.exists() && fileMain.isFile(), "File '$fileMain' must exist"
 
  def numberOfCyclesMain = new XmlSlurper().parse(fileMain).Cycles.Package.size()
  assert numberOfCyclesMain == 0, """We have detected $numberOfCyclesMain cycles. Checkout report file: $fileMain.
                                    """
  File fileTest = new File(jdepend.reportsDir, "test.xml");
  if( fileTest.exists()) {
    def numberOfCyclesTest = new XmlSlurper().parse(fileTest).Cycles.Package.size()
    assert numberOfCyclesTest == 0, """We have detected $numberOfCyclesTest cycles. Checkout report file: $fileTest.
                                    """
  }
}
``` 
- Do you think running `javadoc, javadocJar, sourcesJar, pmd, cpd, checkstyle, spotBugs` by default a good idea? It takes a lot of time when code is modified. I usually and always run them just before commit and release. Is it a bad idea to create a flag to run each task individually? Another to trigger all of checks?
- Where all javadoc build is created? We usually reference it [here](http://javadoc.io/doc/org.joinfaces/joinfaces-build).
- What about creating test-jar and test-source-jar files? They are released to maven today.
- Should we deploy using the following command:
```bash
gradle publish -PdeployRepoUsername=myuser -PdeployRepoPassword=mypass
```

larsgrefer commented:
Some answers:

- The `joinfaces-dependencies.pom` isn't a subproject anymore but a publication of the root project
- Yes, we can extract the versions into properties. I have not done this, because IMO the `dependencyManagement`-block in the root `build.gradle` is centralized enough.
- The duplicated dependency versions across different `build.gradle` files are an error. All Library versions should be defined in the `dependencyManagement`-block of the root `build.gradle` (except omnifaces, which can't be handled that way, because we need two versions of the same artifact here)
- Checks for cpd, duplicated classes and cycles aren't implemented yet, I'll look into that.
- Gradle just runs the tasks it has to. If you don't want all the tasks to run, just invoke the tasks you need.
- An aggregated Javadoc is currently missing, I'll add that.
- We can add the `tests`, `test-sources` and `test-javadoc` if needed, but I don't know a use case for them.
- The credentials for publishing are currently read from the environment variables `OSSRH_USER` and `OSSRH_PASS`, but we can change that.

persapiens commented:
So, let's resume final tasks (I'm ok with other answers):
- Remove duplicated dependency versions across different build.gradle.
- Checks for cpd, duplicated classes and cycles.
- Add aggregated Javadoc.

Just curious:
- What do you think about gradle build written with kotlin? Faster? Easier to maintain?

larsgrefer commented:
- All version numbers are now in the root `build.gradle`-file (except `omnifaces`)
- I've added `nebula.lint`, `cpd` and `jdepend`, but `jdepend` support will be deprecated with Gradle 5 and removed in Gradle 6 because `jdepend` itself isn't maintained anymore.
- The aggregated Javadoc will be published with groupId `org.joinfaces`, artifactId `joinfaces`, classifier `javadoc`

I haven't tried _Kotlin_ yet.
